@gemini_code_assist

```
user@win UCRT64 ~/git/fsearch
$ meson compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\Users\user\Downloads\msys64\ucrt64\bin/ninja.EXE -C C:/Users/user/Downloads/msys64/home/user/git/fsearch/build
ninja: Entering directory `C:/Users/user/Downloads/msys64/home/user/git/fsearch/build'
[1/1] Merging translations for data/io.github.cboxdoerfer.FSearch.metainfo.xml
FAILED: data/io.github.cboxdoerfer.FSearch.metainfo.xml
"C:/Users/user/Downloads/msys64/ucrt64/bin/meson" "--internal" "msgfmthelper" "--msgfmt=C:/Users/user/Downloads/msys64/usr/bin/msgfmt.EXE" "../data/io.github.cboxdoerfer.FSearch.metainfo.xml.in" "data/io.github.cboxdoerfer.FSearch.metainfo.xml" "xml" "C:/Users/user/Downloads/msys64/home/user/git/fsearch/po"
/usr/bin/msgfmt: ITS-Regeln f├╝r ../data/io.github.cboxdoerfer.FSearch.metainfo.xml.in k├Ânnen nicht gefunden werden
ninja: build stopped: subcommand failed.

user@win UCRT64 ~/git/fsearch
$ meson compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\Users\user\Downloads\msys64\ucrt64\bin/ninja.EXE -C C:/Users/user/Downloads/msys64/home/user/git/fsearch/build
ninja: Entering directory `C:/Users/user/Downloads/msys64/home/user/git/fsearch/build'
[1/1] Merging translations for data/io.github.cboxdoerfer.FSearch.metainfo.xml
FAILED: data/io.github.cboxdoerfer.FSearch.metainfo.xml
"C:/Users/user/Downloads/msys64/ucrt64/bin/meson" "--internal" "msgfmthelper" "--msgfmt=C:/Users/user/Downloads/msys64/usr/bin/msgfmt.EXE" "../data/io.github.cboxdoerfer.FSearch.metainfo.xml.in" "data/io.github.cboxdoerfer.FSearch.metainfo.xml" "xml" "C:/Users/user/Downloads/msys64/home/user/git/fsearch/po"
/usr/bin/msgfmt: ITS-Regeln f├╝r ../data/io.github.cboxdoerfer.FSearch.metainfo.xml.in k├Ânnen nicht gefunden werden
ninja: build stopped: subcommand failed.

user@win UCRT64 ~/git/fsearch
$ ls
build  CONTRIBUTING.md  copr  data  debian  fsearch.doap  help  LICENSE  meson.build  meson_options.txt  NEWS  po  README.md  snap  src  TODO.md

user@win UCRT64 ~/git/fsearch
$ cat meson.build
project('fsearch', 'c',
    version: '0.3.alpha0',
    meson_version: '>= 0.45.0',
    default_options: ['c_std=gnu11',
                      'channel=other',
                      ],
    )
app_id = 'io.github.cboxdoerfer.FSearch'

cc = meson.get_compiler('c')

gnome = import('gnome')
i18n = import('i18n')

have_malloc_trim = meson.get_compiler('c').has_function('malloc_trim')

config_h = configuration_data()
config_h.set('HAVE_MALLOC_TRIM', have_malloc_trim)
config_h.set_quoted('APP_ID', app_id)
config_h.set_quoted('PACKAGE_VERSION', meson.project_version())
config_h.set_quoted('VERSION', meson.project_version())
config_h.set_quoted('GETTEXT_PACKAGE', 'fsearch')
config_h.set_quoted('LOCALEDIR', join_paths(get_option('prefix'), get_option('localedir')))
config_h.set_quoted('PACKAGE_WEBSITE', 'https://github.com/cboxdoerfer/fsearch')
config_h.set_quoted('PACKAGE_ICON_NAME', app_id)
config_h.set_quoted('PACKAGE_NAME', 'FSearch')
config_h.set_quoted('BUILD_CHANNEL', get_option('channel'))

# ensure off_t is 64bit
add_project_arguments('-DHAVE_CONFIG_H', '-D_FILE_OFFSET_BITS=64', '-D_GNU_SOURCE', language : 'c')

fsearch_include_dirs = include_directories(
  '.',
)

subdir('data')
subdir('help')
subdir('src')
subdir('po')

configure_file(
    output: 'config.h',
    configuration: config_h,
    )

```

--- a/src/fsearch_config.c
+++ b/src/fsearch_config.c
@@ -249,7 +249,7 @@ config_load(FsearchConfig *config) {
 
         // Applications
 #ifdef _WIN32
-        config->folder_open_cmd = config_load_string(key_file, "Applications", "folder_open_cmd", "explorer \"%p\"");
+        config->folder_open_cmd = config_load_string(key_file, "Applications", "folder_open_cmd", "explorer.exe \"{path_full_raw}\"");
 #else
         config->folder_open_cmd = config_load_string(key_file, "Applications", "folder_open_cmd", NULL);
 #endif

--- a/src/fsearch_file_utils.c
+++ b/src/fsearch_file_utils.c
@@ -292,6 +292,8 @@ create_uris_launch_context(const char *content_type, GPtrArray *files, FsearchFi
     }
 
     GAppInfo *app_info = g_app_info_get_default_for_type(content_type, FALSE);
+    g_debug("[open_files] Looking for default app for content type: %s, found: %s", 
+            content_type, app_info ? "yes" : "no");
     if (!app_info) {
 #ifdef _WIN32
         // On Windows, if no default application is registered for directory content types,
@@ -326,15 +328,20 @@ create_uris_launch_context(const char *content_type, GPtrArray *files, FsearchFi
             for (uint32_t i = 0; i < files->len; ++i) {
                 GFile *file = g_ptr_array_index(files, i);
                 g_autofree char *uri = g_file_get_uri(file);
+                g_autofree char *path = g_file_get_path(file);
                 if (uri) {
+                    g_debug("[Windows] Attempting to open file with system default: %s (URI: %s, Content-Type: %s)", 
+                            path ? path : "unknown", uri, content_type);
                     g_autoptr(GError) error = NULL;
                     if (!g_app_info_launch_default_for_uri(uri, ctx->app_launch_context, &error)) {
-                        g_autofree char *path = g_file_get_path(file);
+                        g_debug("[Windows] Failed to open file with system default: %s", error ? error->message : "unknown error");
                         add_error_message_with_format(ctx->error_messages,
                                                       C_("Will be followed by the path of the file.",
                                                          "Error while opening file"),
                                                       path ? path : uri,
                                                       error->message);
+                    } else {
+                        g_debug("[Windows] Successfully launched system default for file: %s", path ? path : uri);
                     }
                 }
             }
@@ -354,10 +361,16 @@ create_uris_launch_context(const char *content_type, GPtrArray *files, FsearchFi
     FsearchFileUtilsLaunchUrisContext *launch_uris_ctx = g_new0(FsearchFileUtilsLaunchUrisContext, 1);
     launch_uris_ctx->app_info = g_steal_pointer(&app_info);
 
+    g_debug("[open_files] Using app_info for content type %s: %s", content_type, 
+            g_app_info_get_name(launch_uris_ctx->app_info));
+
     for (uint32_t i = 0; i < files->len; ++i) {
         GFile *file = g_ptr_array_index(files, i);
         char *uri = g_file_get_uri(file);
         if (uri) {
+            g_autofree char *path = g_file_get_path(file);
+            g_debug("[open_files] Adding file to launch queue: %s (URI: %s)", 
+                    path ? path : "unknown", uri);
             launch_uris_ctx->uris = g_list_append(launch_uris_ctx->uris, uri);
         }
     }

--- a/src/fsearch_file_utils.c
+++ b/src/fsearch_file_utils.c
@@ -140,14 +140,17 @@ build_folder_open_cmd(const char *path, const char *path_full, const char *cmd) 
     
     // Also support legacy %p token for backward compatibility
     if (result && strstr(result, "%p")) {
+        g_autoptr(GRegex) legacy_regex = g_regex_new("%p", 0, 0, NULL);
+        if (legacy_regex) {
 #ifdef _WIN32
-        // On Windows, use the raw path without additional quoting since %p was already quoted in the template
-        g_autofree char *temp = g_strdup(result);
-        result = g_strreplace(temp, "%p", path_full, -1);
+            // On Windows, use the raw path without additional quoting since %p was already quoted in the template
+            g_autofree char *temp = g_steal_pointer(&result);
+            result = g_regex_replace_literal(legacy_regex, temp, -1, 0, path_full, 0, NULL);
 #else
-        g_autofree char *temp = g_strdup(result);
-        result = g_strreplace(temp, "%p", path_full_quoted, -1);
+            g_autofree char *temp = g_steal_pointer(&result);
+            result = g_regex_replace_literal(legacy_regex, temp, -1, 0, path_full_quoted, 0, NULL);
 #endif
+        }
     }
     
     return g_steal_pointer(&result);

--- a/src/fsearch_file_utils.c
+++ b/src/fsearch_file_utils.c
@@ -100,8 +100,15 @@ build_folder_open_cmd(const char *path, const char *path_full, const char *cmd) 
     if (!path || !path_full) {
         return NULL;
     }
+    
+#ifdef _WIN32
+    // On Windows, use simple double quotes for paths instead of shell quoting
+    g_autofree char *path_quoted = g_strdup_printf("\"%s\"", path);
+    g_autofree char *path_full_quoted = g_strdup_printf("\"%s\"", path_full);
+#else
     g_autofree char *path_quoted = g_shell_quote(path);
     g_autofree char *path_full_quoted = g_shell_quote(path_full);
+#endif
 
     // The following code is mostly based on the example code found here:
     // https://developer.gnome.org/glib/stable/glib-Perl-compatible-regular-expressions.html#g-regex-replace-eval
@@ -129,7 +136,21 @@ build_folder_open_cmd(const char *path, const char *path_full, const char *cmd) 
     // surrounded with {}
     g_autoptr(GRegex) reg = g_regex_new("{[\\w]+}", 0, 0, NULL);
     // Replace all the matched keywords
-    return g_regex_replace_eval(reg, cmd, -1, 0, 0, keyword_eval_cb, keywords, NULL);
+    g_autofree char *result = g_regex_replace_eval(reg, cmd, -1, 0, 0, keyword_eval_cb, keywords, NULL);
+    
+    // Also support legacy %p token for backward compatibility
+    if (result && strstr(result, "%p")) {
+#ifdef _WIN32
+        // On Windows, use the raw path without additional quoting since %p was already quoted in the template
+        g_autofree char *temp = g_strdup(result);
+        result = g_strreplace(temp, "%p", path_full, -1);
+#else
+        g_autofree char *temp = g_strdup(result);
+        result = g_strreplace(temp, "%p", path_full_quoted, -1);
+#endif
+    }
+    
+    return g_steal_pointer(&result);
 }
 
 static bool
@@ -271,12 +292,50 @@ create_uris_launch_context(const char *content_type, GPtrArray *files, FsearchFi
     if (!app_info) {
 #ifdef _WIN32
         // On Windows, if no default application is registered for directory content types,
-        // fall back to using explorer.exe directly
+        // we need to handle directories specially by launching explorer.exe with the path
         if (strcmp(content_type, "application/x-directory") == 0 || strcmp(content_type, "inode/directory") == 0) {
-            app_info = g_app_info_create_from_commandline("explorer.exe", "File Explorer", G_APP_INFO_CREATE_NONE, NULL);
-            if (app_info) {
-                g_debug("[Windows] Using explorer.exe fallback for directory content type: %s", content_type);
+            // For directories on Windows, we launch explorer.exe individually for each path
+            // instead of trying to batch them, since explorer.exe needs the specific path
+            for (uint32_t i = 0; i < files->len; ++i) {
+                GFile *file = g_ptr_array_index(files, i);
+                g_autofree char *path = g_file_get_path(file);
+                if (path) {
+                    g_autofree char *quoted_path = g_shell_quote(path);
+                    g_autofree char *command = g_strdup_printf("explorer.exe %s", quoted_path);
+                    g_autoptr(GError) error = NULL;
+                    if (!g_spawn_command_line_async(command, &error)) {
+                        add_error_message_with_format(ctx->error_messages,
+                                                      C_("Will be followed by the path of the folder.",
+                                                         "Error while opening folder"),
+                                                      path,
+                                                      error->message);
+                    } else {
+                        g_debug("[Windows] Launched explorer.exe for directory: %s", path);
+                    }
+                }
             }
+            return;
+        }
+        // For non-directory files on Windows, try to use the system default
+        // This should work for most file types including images
+        else {
+            // Use the system's default handler for this file type
+            for (uint32_t i = 0; i < files->len; ++i) {
+                GFile *file = g_ptr_array_index(files, i);
+                g_autofree char *uri = g_file_get_uri(file);
+                if (uri) {
+                    g_autoptr(GError) error = NULL;
+                    if (!g_app_info_launch_default_for_uri(uri, ctx->app_launch_context, &error)) {
+                        g_autofree char *path = g_file_get_path(file);
+                        add_error_message_with_format(ctx->error_messages,
+                                                      C_("Will be followed by the path of the file.",
+                                                         "Error while opening file"),
+                                                      path ? path : uri,
+                                                      error->message);
+                    }
+                }
+            }
+            return;
         }
 #endif
         if (!app_info) {

